caravel/models.py

```diff
@@ -1667,3 +1672,48 @@ class FavStar(Model):
     class_name = Column(String(50))
     obj_id = Column(Integer)
     dttm = Column(DateTime, default=func.now())
+
+
+class SqlExprMacro(Model, AuditMixinNullable):
+
+    """SQL expression macros based on Jinja2 template"""
+
+    __tablename__ = 'sql_expr_macros'
+
+    id = Column(Integer, primary_key=True)
+    namespace = Column(String(50))
+    source_code = Column(Text)
+    _env = None
+    _compiled = False
+
+    @classmethod
+    def recompile(cls):
+        """Re-compile all macros into a Jinja2 environment"""
+        macros = db.session.query(cls).all()
+        if macros:
+            mapping = {m.namespace: m.source_code for m in macros}
+            env = jinja2.Environment(loader=jinja2.DictLoader(mapping))
+            for namespace in mapping:
+                env.globals[namespace] = env.get_template(namespace).module
+            cls._env = env
+        else:
+            cls._env = None
+        cls._compiled = True
```

Comment:
I think some of these operations are not thread-safe, but I think it's fine since we use gunicorn with independent single-threaded workers.